### ISSUE TITLE
[Bug]: ComboBox downshiftProps missing reference to Downshift component where state change can be done

### ISSUE THREAD

**Issue by ashishkrz** [carbon-design-system/carbon#17027]
### Package

@carbon/react

### Browser

Chrome

### Package version

1.62.1

### React version

17.0.2

### Description

ComboBox has downshiftProps where we we used to utilize onStateChange function to make any change directly to Downshift Component by changing the state. Previously onStateChange used to receive Downshift component as 2nd argument, but now there is no way to receive the component, it is always undefined.

### Reproduction/example

https://stackblitz.com/edit/github-v56j87?file=src%2FApp.jsx

### Steps to reproduce

1. Select one Item from ComboBox
2. Now try to clear the selection using close icon, it is not getting cleared.
3. Now if you try to open ComboBox and now try doing clear, it will get cleared.

So basically the issue is, when item is selected and ComboBox is closed (not expanded), then item not getting cleared when you click close icon.

### Suggested Severity

Severity 2 = User cannot complete task, and/or no workaround within the user experience of a given component.

### Application/PAL

Power Virtual Server

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/carbon/blob/f555616971a03fd454c0f4daea184adf41fff05b/.github/CODE_OF_CONDUCT.md)
- [X] I checked the [current issues](https://github.com/carbon-design-system/carbon/issues) for duplicate problems

**Comment by guidari** [comment]
Just for context, in `v1.57.0` that was working.

**Comment by ashishkrz** [comment]
> Just for context, in `v1.57.0` that was working.

Checked just now, Yes it works on v1.57.0. But here also on 1st click it's not getting cleared sometimes. 2nd click onwards it works fine.

**Comment by ali-sadeghin** [comment]
I confirm too: it worked in 1.57.0 and totally broke in our UI from 1.58.0

**Comment by tay1orjones** [comment]
I believe this is due to an incompatibility between `useCombobox` and `<Downshift>`. The [`<Downshift>` signature](https://github.com/downshift-js/downshift?tab=readme-ov-file#onstatechange) is 
```
function(changes: object, stateAndHelpers: object)
```

The [useCombobox signature](https://github.com/downshift-js/downshift/tree/master/src/hooks/useCombobox#onstatechange) is
```
function(changes: object)
``` 

It [sounds like](https://github.com/downshift-js/downshift/pull/812#:~:text=Doesn%27t%20have%20goodies%20such%20as%20stateAndHelpers) this is a known/intentional incompatibility.

I believe the update to use `useCombobox` may have shipped in 1.58.0.  I'm not sure if there is a workaround here. In the stackblitz you're accessing the input value, could you get the same via a `ref`? It's placed internally directly on the `input`

https://github.com/carbon-design-system/carbon/blob/8fe1ee3a1522371235dd15e481770adeea5d6d08/packages/react/src/components/ComboBox/ComboBox.tsx#L690-L700

**Comment by ashishkrz** [comment]
@tay1orjones Could you please help on stackblitz, just by modifying things using ref like you suggested?

**Comment by clouria** [comment]
I am also running into a similar issue to this. I am unable to use the `onChange` and `stateReducer` in the `downshiftProps ` for `<ComboBox />` which were working prior to v1.58.0 and seem to be related to the `useCombobox` change.

**Comment by mateBe95** [comment]
bump up, its break main functionalities in our product which is broken. Please provide the fix @tay1orjones in the new version.

**Comment by ashishkrz** [comment]
@tay1orjones Just to confirm what is changed now? I see that the position of spreaded downshiftProps got changed but that will not fix this issue, I believe. Still the 2nd argument to onStateChange function remains undefined.

Could you please help over stackblitz https://stackblitz.com/edit/github-v56j87?file=src%2FApp.jsx making it work with the fix done?

**Comment by ali-sadeghin** [comment]
@tay1orjones I tried 1.63.1 and unfortunately, it doesn't help. Still broken.

**Comment by tay1orjones** [comment]
@ashishkrz @ali-sadeghin Thanks for validating this and letting us know. I see the fix only covers half the problem. Your `downshiftProps` now fully override any used within Combobox, but the other missing half (if I'm understanding correctly) is enabling you to modify downshift's internal state.

I think I've found a way to patch that gap. It may require some changes on your end, but they should be minor. I should have a PR up today and we can publish a patch release for it as soon as it's merged.

Sorry for the delay and I appreciate your patience here. How `downshiftProps` is used is opaque to us outside of reports like these. It's hard to know when or where we need to patch the downshift APIs as it shifts over time.

**Comment by ali-sadeghin** [comment]
@tay1orjones is the fix planned to be built any time soon? We have a comming push to prod ...
Thanks!

**Comment by guidari** [comment]
Hey @ali-sadeghin 
It was just release in the [v1.63.2](https://www.npmjs.com/package/@carbon/react?activeTab=versions)

**Comment by ashishkrz** [comment]
@tay1orjones @guidari  I tried the version 1.63.2, it's still broken. Could you please check? Here is the stackblitz [link](https://stackblitz.com/edit/github-v56j87?file=src%2FApp.jsx) for your ref which is using 1.63.2 but still it is not working.

**Comment by ashishkrz** [comment]
@tay1orjones The behaviour is still similar, it's not working. Could you please help me on the same if I need to use it differently? @ali-sadeghin How it worked for you, I have created a stackblitz (https://stackblitz.com/edit/github-v56j87?file=src%2FApp.jsx) which is using latest version of carbon, could you please give it a try there and check once? 

Our PowerVS team is constantly looking for the fix for the same, will be helpful if you update on same. Thanks!

**Comment by tay1orjones** [comment]
Sorry for the confusion, details were mentioned on the follow up PR:

> If a consumer was using an action that is no longer available, such as `setState`, a functional equivalent can usually be found. This will require a code change in consuming projects.
> 
> ```diff
> - stateAndHelpers.setState({ inputValue: 'Item 1' });
> + downshiftActions.current.setInputValue('Item 1');
> ```

The PR also contained an [updated version](https://github.com/carbon-design-system/carbon/pull/17118/files#diff-ebf550693a4776b9e150a27234bfc05cb4944224ab2cfcefde4005cd5e32f47e) of your same example code you can reference:

**Comment by tay1orjones** [comment]
@ashishkrz Sorry for the confusion, details were mentioned on the follow up PR:

> If a consumer was using an action that is no longer available, such as `setState`, a functional equivalent can usually be found. This will require a code change in consuming projects.
> 
> ```diff
> - stateAndHelpers.setState({ inputValue: 'Item 1' });
> + downshiftActions.current.setInputValue('Item 1');
> ```

The PR also contained an [updated version](https://github.com/carbon-design-system/carbon/pull/17118/files#diff-ebf550693a4776b9e150a27234bfc05cb4944224ab2cfcefde4005cd5e32f47e) of your same example code you can reference:

**Comment by ashishkrz** [comment]
@tay1orjones Selection works fine but onChange is not getting called after the item is selected from combobox. Could you please check once?

**Comment by ashishkrz** [comment]
@tay1orjones I have updated the code on stackblitz https://stackblitz.com/edit/github-v56j87?file=src%2FApp.jsx for you reference to check it quickly. 

The onChange is not getting called after a selection:

`onChange={(e) => {
          console.info('HELLO', e);
        }}`

**Comment by tay1orjones** [comment]
This should be fixed in 1.72.0 - if you're still seeing a problem with this, please open a new issue. Thanks!

**Comment by ashishkrz** [comment]
@tay1orjones Created issue for same https://github.com/carbon-design-system/carbon/issues/18212. Please take a look.
